test/websocket/test_websocket_send.c

```diff
@@ -15,43 +21,50 @@ EM_BOOL WebSocketOpen(int eventType, const EmscriptenWebSocketOpenEvent *e, void
 }
 
 EM_BOOL WebSocketClose(int eventType, const EmscriptenWebSocketCloseEvent *e, void *userData) {
-  printf("close(eventType=%d, wasClean=%d, code=%d, reason=%s, userData=%p)\n", eventType, e->wasClean, e->code, e->reason, userData);
+  printf("close(socket=%d, eventType=%d, wasClean=%d, code=%d, reason=%s, userData=%p)\n", e->socket, eventType, e->wasClean, e->code, e->reason, userData);
   return 0;
 }
 
 EM_BOOL WebSocketError(int eventType, const EmscriptenWebSocketErrorEvent *e, void *userData) {
-  printf("error(eventType=%d, userData=%p)\n", eventType, userData);
+  printf("error(socket=%d, eventType=%d, userData=%p)\n", e->socket, eventType, userData);
   return 0;
 }
 
 EM_BOOL WebSocketMessage(int eventType, const EmscriptenWebSocketMessageEvent *e, void *userData) {
-  printf("message(eventType=%d, userData=%p data=%p, numBytes=%d, isText=%d)\n", eventType, userData, e->data, e->numBytes, e->isText);
+  printf("message(socket=%d, eventType=%d, userData=%p data=%p, numBytes=%d, isText=%d)\n", e->socket, eventType, userData, e->data, e->numBytes, e->isText);
   static int text_received = 0;
+  static int binary_received = 0;
+  assert(e->socket == sock1 || e->socket == sock2);
   if (e->isText) {
     printf("text data: \"%s\"\n", e->data);
     assert(strcmp((const char*)e->data, "hello on the other side") == 0);
-    text_received = 1;
-    return 0;
-  }
-
-  // We expect to receive the text message beofre the binary one
-  assert(text_received);
-  printf("binary data:");
-  for (int i = 0; i < e->numBytes; ++i) {
-    printf(" %02X", e->data[i]);
-    assert(e->data[i] == i);
+    text_received++;
+  } else {
+    // We expect to receive the text message beofre the binary one
+    printf("binary data:");
+    for (int i = 0; i < e->numBytes; ++i) {
+      printf(" %02X", e->data[i]);
+      assert(e->data[i] == i);
+    }
+    printf("\n");
+    binary_received++;
+    emscripten_websocket_close(e->socket, 0, 0);
+    emscripten_websocket_delete(e->socket);
+    if (e->socket == sock1) {
+      sock1 = 0;
+    } else {
+      sock2 = 0;
+    }
+    // Once both sockets are closed we are done.
+    if (!sock1 && !sock2) {
+      emscripten_force_exit(0);
+    }
```

Comment:
This logic should be outside of the else, no? That is the only way that we can accept both orders, I think, and not depend on binary being last.

Reply:
Whatever interleaving happens the binary message will always be last.   Since for each socket the binary message comes second.

We only close the socket once we get the binary message from it, so the only place that sock1 or sock2 could be made to be null is here inside this else branch.

Reply:
In that case we can put a return after line 39, and avoid indenting all the rest in the `else`, the way it was before? Or am I missing a reason to put this code in the indented position?

Reply:
lgtm with that change + the typo in the other comment.

Reply:
Why do you prefer adding an early return and putting the binary case outside the `if/else`?   The `if/else` seems quite natural to be since it one case is `text` and other other is `binary`?

Reply:
I guess it makes the diff smaller?   Done.

Reply:
Yeah, it is more symmetrical the other way, but in general I think an early return is nice when one arm is much longer, as it allows that arm to be unindented. But I was also unsure if the indentation had another reason here, that's why I was asking (if we allowed more orders than before). Anyhow, lgtm either way.